hazelcast/src/main/java/com/hazelcast/config/MerkleTreeConfig.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.annotation.Beta;
+import com.hazelcast.util.StringUtil;
+
+import java.io.IOException;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+/**
+ * Configuration for a merkle tree.
+ * The merkle tree is a data structure used for efficient comparison of the
+ * difference in the contents of large data structures. The precision of
+ * such a comparison mechanism is defined by the depth of the merkle tree.
+ * <p>
+ * A larger depth means that a data synchronization mechanism will be able
+ * to pinpoint a smaller subset of the data structure contents in which a
+ * change occurred. This causes the synchronization mechanism to be more
+ * efficient. On the other hand, a larger tree depth means the merkle tree
+ * will consume more memory.
+ * <p>
+ * A smaller depth means the data synchronization mechanism will have to
+ * transfer larger chunks of the data structure in which a possible change
+ * happened. On the other hand, a shallower tree consumes less memory.
+ * The depth must be between {@value MIN_DEPTH} and {@value MAX_DEPTH}
+ * (exclusive). The default depth is {@value DEFAULT_DEPTH}.
+ * <p>
+ * As the comparison mechanism is iterative, a larger depth will also prolong
+ * the duration of the comparison mechanism. Care must be taken to not have
+ * large tree depths if the latency of the comparison operation is high.
+ * The default depth is {@value DEFAULT_DEPTH}.
+ * <p>
+ * This configuration is not tied to a specific data structure and can be
+ * reused.
+ * <p>
+ * See https://en.wikipedia.org/wiki/Merkle_tree.
+ *
+ * @since 3.11
+ */
+@Beta
+public class MerkleTreeConfig implements IdentifiedDataSerializable {
+    /**
+     * Minimal depth of the merkle tree.
+     */
+    private static final int MIN_DEPTH = 2;
+    /**
+     * Maximal depth of the merkle tree.
+     */
+    private static final int MAX_DEPTH = 27;
+    /**
+     * Default depth of the merkle tree.
+     */
+    private static final int DEFAULT_DEPTH = 10;
+
+    private boolean enabled = true;
+    private int depth = DEFAULT_DEPTH;
+    private String mapName;
```

Comment:
seems this should be something like `dataStructureName` instead of `mapName`

Reply:
Ah, now I remember. This is supposed to be the same way as `EventJournalConfig`. If we put `dataStructureName`, we now know it applies to maps.
If we add merkle trees to ICache, `dataStructureName` is ambiguous. `EventJournalConfig` avoids this by having:
```
private String mapName;
    private String cacheName;
```

Reply:
I'll remove the javadoc.